### PR TITLE
[WIP] Expand variadic functions in IR

### PR DESCRIPTION
**Pull request by JonChesterfield**
Rewrite calls to variadic functions into calls to an equivalent non-variadic function.

This makes calls to known variadic functions a zero cost abstraction. The GPUs use it as a backend implementation.

The pass runs for implemented targets on known functions at O1 or above. For targets using it as codegen it runs from the backend on known and unknown functions. Interacts well with inlining so is scheduled shortly before the inliner.

Relative to the abandoned pull/81058, this is the whole end to end transform as opposed to a subset for easier review. The feedback on that is applied, in particular this makes no attempt to detect existing va_list functions, so in some cases it'll make the IR worse in a fashion that the inliner completely reverts. The ABI abstraction is reworked to be more orthogonal which hopefully makes it clear that the switch isn't worth scattering across the target classes.

Marked WIP because vector types in structs on X64 might be misbehaving. Still testing this - it's solid enough to build a working clang and libxml2 and should meet the use cases in gpu libc and I'll put a patch up for the rocm CI - but would like more test coverage and probably another architecture implemented before landing.

### CODE REVIEW

File: clang/lib/CodeGen/ABIInfoImpl.cpp [hunk 1]
```diff
@@ -154,11 +154,20 @@ llvm::Value *CodeGen::emitRoundPointerUpToAlignment(CodeGenFunction &CGF,
                                                     llvm::Value *Ptr,
                                                     CharUnits Align) {
   // OverflowArgArea = (OverflowArgArea + Align - 1) & -Align;
+  Ptr = CGF.Builder.CreateAddrSpaceCast(Ptr, CGF.AllocaInt8PtrTy,
+                                        Ptr->getName() + ".addrcast");
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/ABIInfoImpl.cpp:158]: I can sort of guess why you're doing this, but probably you want to rename the method so it's clear why it's specialized this way.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+
+    CallInst::TailCallKind TCK = CI->getTailCallKind();
+    assert(TCK != CallInst::TCK_MustTail); // guarded at prologue
+
+    // It doesn't get to be a tail call any more
+    // might want to guard this with arch, x64 and aarch64 document that
+    // varargs can't be tail called anyway
+    // Not totally convinced this is necessary but dead store elimination
+    // will discard the stores to the Alloca and pass uninitialised memory along
+    // instead when the function is marked tailcall
```
Comment (efriedma-quic) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:957]: I'll take another shot at explaining this...

From LangRef: "[tail and musttail] imply that the callee does not access allocas from the caller."  This applies whether or not it's actually possible to do sibling-call/tail-call optimization on the call.  And in fact, "tail" has basically no other semantic implications.  So LLVM IR optimizations infer "tail" pretty aggressively.

Because of this definition, alias analysis assumes calls marked "tail" don't read/write allocas. So the DSE behavior is expected.

File: clang/lib/CodeGen/ABIInfoImpl.cpp (hunk 1 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/ABIInfoImpl.cpp:158]: This might not be the right place for it - amdgpu passes a pointer in addrspace(5) which then hits an assert with the current implementation. The va_* intrinsics specify an address space now which should permit a cleaner answer, though I'm seeing p0 used from clang for targets where it shouldn't be. That's on the WIP part.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 2 above)
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:957]: Tail on varargs functions is sort of valid in that the callee doesn't access an alloca. The callee does tend to dig around in the stack of the caller though, so tail calling varargs functions only works out nicely if the callee rearranges the stack frame on behalf of the caller. Which it can do, but usually doesn't.

I think your reading is right that DSE making this inference is reasonable. I think it's also right that a varargs function could be tail called on some implementations, though isn't on the ones considered so far in this patch.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+}
+
+PreservedAnalyses ExpandVariadicsPass::run(Module &M, ModuleAnalysisManager &) {
+  return ExpandVariadics(ConstructedMode).runOnModule(M)
+             ? PreservedAnalyses::none()
+             : PreservedAnalyses::all();
+}
+
+ExpandVariadicsPass::ExpandVariadicsPass(OptimizationLevel Level)
+    : ExpandVariadicsPass(Level == OptimizationLevel::O0
```
Comment (jhuber6) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:1051]: I don't think this works, there's a cyclic dependency here on `LLVMPasses` which defines this. The standard way to handle this should be when we register the pass in the `PassBuilderPipeline`.

File: clang/lib/CodeGen/ABIInfoImpl.cpp [hunk 4]
```diff
@@ -154,11 +154,20 @@ llvm::Value *CodeGen::emitRoundPointerUpToAlignment(CodeGenFunction &CGF,
...
+  // ptrmask is sensitive to the bitwidth of the mask
+  unsigned IndexTypeSize =
+      CGF.CGM.getDataLayout().getIndexTypeSizeInBits(RoundUp->getType());
+  llvm::IntegerType *MaskType =
+      llvm::IntegerType::get(CGF.getLLVMContext(), IndexTypeSize);
+
   return CGF.Builder.CreateIntrinsic(
-      llvm::Intrinsic::ptrmask, {CGF.AllocaInt8PtrTy, CGF.IntPtrTy},
-      {RoundUp, llvm::ConstantInt::get(CGF.IntPtrTy, -Align.getQuantity())},
+      llvm::Intrinsic::ptrmask, {CGF.AllocaInt8PtrTy, MaskType},
```
Comment (arsenm) [inline clang/lib/CodeGen/ABIInfoImpl.cpp:169]: This seems like it was bad to begin with, I think this should have just been using Ptr->getType() instead of the CFG types

File: clang/lib/CodeGen/Targets/AMDGPU.cpp [hunk 5]
```diff
@@ -115,7 +115,13 @@ void AMDGPUABIInfo::computeInfo(CGFunctionInfo &FI) const {
 
 Address AMDGPUABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
                                  QualType Ty) const {
-  llvm_unreachable("AMDGPU does not support varargs");
+  const bool IsIndirect = false;
+  const bool AllowHigherAlign = true;
+  // Would rather not naturally align values
+  // Splitting {char, short} into two separate arguments makes that difficult.
```
Comment (arsenm) [inline clang/lib/CodeGen/Targets/AMDGPU.cpp:121]: Not sure what this comment is saying. 4 is the only correct alignment for anything stack related

File: llvm/include/llvm/Transforms/IPO/ExpandVariadics.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#include "llvm/IR/PassManager.h"
+
+namespace llvm {
+
+class Module;
+class ModulePass;
+class OptimizationLevel;
+
+enum class ExpandVariadicsMode {
+  unspecified,
```
Comment (arsenm) [inline llvm/include/llvm/Transforms/IPO/ExpandVariadics.h:20]: Capitalize and document the fields

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 7]
```diff
@@ -0,0 +1,1056 @@
...
+
+namespace {
+
+// Module implements getFunction() which returns nullptr on missing declaration
+// and getOrInsertFunction which creates one when absent. Intrinsics.h
+// implements getDeclaration which creates one when missing. This should be
+// changed to be consistent with Module()'s naming. Implementing as a local
+// function here in the meantime to decouple from that process.
+Function *getPreexistingDeclaration(Module *M, Intrinsic::ID id,
+                                    ArrayRef<Type *> Tys = std::nullopt) {
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:82]:
```suggestion
                                    ArrayRef<Type *> Tys = {}) {
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 8]
```diff
@@ -0,0 +1,1056 @@
...
+    };
+
+    Idxs[2] = ConstantInt::get(I32, 0);
+    Builder.CreateStore(
+        ConstantInt::get(I32, 48),
+        Builder.CreateInBoundsGEP(VaListTy, VaList, Idxs, "gp_offset"));
+
+    Idxs[2] = ConstantInt::get(I32, 1);
+    Builder.CreateStore(
+        ConstantInt::get(I32, 6 * 8 + 8 * 16),
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:191]: Comment on where these magic expressions were derived?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 9]
```diff
@@ -0,0 +1,1056 @@
...
+        Builder.CreateInBoundsGEP(VaListTy, VaList, Idxs, "fp_offset"));
+
+    Idxs[2] = ConstantInt::get(I32, 2);
+    Builder.CreateStore(
+        VoidBuffer,
+        Builder.CreateInBoundsGEP(VaListTy, VaList, Idxs, "overfow_arg_area"));
+
+    Idxs[2] = ConstantInt::get(I32, 3);
+    Builder.CreateStore(
+        ConstantPointerNull::get(PointerType::getUnqual(Ctx)),
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:201]: Don't use getUnqual, use the explicit address space. Best to query the alloca addrspace?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 10]
```diff
@@ -0,0 +1,1056 @@
...
+    // FIXME: It looks like a module can contain arbitrary integers for address
+    // spaces in which case we might need to check _lots_ of cases. Maybe add a
+    // rule to the verifier that the vastart/vaend intrinsics can have arguments
+    // in 0 or in allocaaddrspace but nowhere else
+    SmallSet<unsigned, 2> Set;
+    Set.insert(0); // things tend to end up in zero
+    Set.insert(
+        DL.getAllocaAddrSpace()); // the argument should be in alloca addrspace
+    return Set;
+  }
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:412]: I don't understand why you would need to track a set of these. These should really end up just using the alloca address space

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 11]
```diff
@@ -0,0 +1,1056 @@
...
+
+    if (!rewriteABI()) {
+      // e.g. can't replace a weak function unless changing the original symbol
+      if (GlobalValue::isInterposableLinkage(F->getLinkage())) {
+        return false;
+      }
+    }
+
+    if (!rewriteABI()) {
+      // If optimising, err on the side of leaving things alone
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:434]: Move all this to an allUsesAreTrivialCalls helper function?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 12]
```diff
@@ -0,0 +1,1056 @@
...
+
+  // At pass input, va_start intrinsics only occur in variadic functions, as
+  // checked by the IR verifier.
+
+  // The lowering pass needs to run on all variadic functions.
+  // The optimise could run on only those that call va_start
+  // in exchange for additional book keeping to avoid transforming
+  // the same function multiple times when it contains multiple va_start.
+  // Leaving that compile time optimisation for a later patch.
+  for (Function &F : llvm::make_early_inc_range(M))
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:584]: Don't need llvm::

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 13]
```diff
@@ -0,0 +1,1056 @@
...
+  // equivalent of a va_copy from the newly appended va_list argument, va_end
+  // and va_copy are removed. All that remains is for the lowering pass to find
+  // indirect calls and rewrite those as well.
+
+  if (Mode == ExpandVariadicsMode::lowering) {
+    for (Function &F : llvm::make_early_inc_range(M)) {
+      if (F.isDeclaration())
+        continue;
+
+      // Now need to track down indirect calls. Can't find those
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:616]: Use InstVisitor? Or do something else about the indentation level?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 14]
```diff
@@ -0,0 +1,1056 @@
...
+
+  if (rewriteABI()) {
+    // No direct calls remain to F, remaining uses are things like address
+    // escaping, modulo errors in this implementation.
+    for (User *U : llvm::make_early_inc_range(F->users()))
+      if (CallBase *CB = dyn_cast<CallBase>(U)) {
+        Value *calledOperand = CB->getCalledOperand();
+        if (F == calledOperand) {
+          report_fatal_error(
+              "ExpandVA abi requires eliminating call uses first\n");
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:677]: Don't think these are supposed to have newlines

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 15]
```diff
@@ -0,0 +1,1056 @@
...
+        }
+      }
+
+    Changed = true;
+    // Converting the original variadic function in-place into the equivalent
+    // one.
+    Equivalent->setLinkage(F->getLinkage());
+    Equivalent->setVisibility(F->getVisibility());
+    Equivalent->takeName(F);
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:686]: Can you use copyAttributesFrom?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 16]
```diff
@@ -0,0 +1,1056 @@
...
+  ArgTypes.push_back(ABI.VAList->vaListParameterType(M));
+
+  FunctionType *NFTy = inlinableVariadicFunctionType(M, F.getFunctionType());
+  Function *NF = Function::Create(NFTy, F.getLinkage(), F.getAddressSpace());
+
+  // Note - same attribute handling as DeadArgumentElimination
+  NF->copyAttributesFrom(&F);
+  NF->setComdat(F.getComdat()); // beware weak
+  F.getParent()->getFunctionList().insert(F.getIterator(), NF);
+  NF->setName(F.getName() + ".valist");
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:724]: If it's ABI the new function should retain the original name?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 17]
```diff
@@ -0,0 +1,1056 @@
...
+
+  // Put the alloca to hold the variadic args in the entry basic block.
+  // The clumsy construction is to set a the alignment on the instance
+  Builder.SetInsertPointPastAllocas(CBF);
+
+  // The struct instance needs to be at least MaxFieldAlign for the alignment of
+  // the fields to be correct at runtime. Use the native stack alignment instead
+  // if that's greater as that tends to give better codegen.
+  Align AllocaAlign = MaxFieldAlign;
+  if (DL.exceedsNaturalStackAlignment(Align(1024))) {
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:890]: Where did 1024 come from?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 18]
```diff
@@ -0,0 +1,1056 @@
...
+  Builder.SetInsertPoint(CB);
+
+  Builder.CreateLifetimeStart(Alloced, sizeOfAlloca(Ctx, DL, Alloced));
+
+  Frame.initialiseStructAlloca(DL, Builder, Alloced);
+
+  unsigned NumArgs = FuncType->getNumParams();
+
+  SmallVector<Value *> Args;
+  Args.assign(CB->arg_begin(), CB->arg_begin() + NumArgs);
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:911]: Move to constructor?

File: clang/lib/CodeGen/Targets/AMDGPU.cpp (hunk 5 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:121]: Currently amdgpu takes a `struct {char x; short y;} example;` and splats it across two registers. Doesn't seem totally good to me but I'm trying to avoid changing the non-variadic calling convention at the same time.

This means foo(example) and bar (char, short) look the same in IR, thus they need to have consistent representations as far as va_arg is concerned. The specific case that caused me trouble here was trying to pass a double with four byte alignment as opposed to eight (that's what the the naturally align values comment relates to).

I'll see if I can capture this in an amdgpu specific test case and fix up the comment

Comment (arsenm) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:121]: I thought all structs under 8 bytes were packed into i32/i64

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 17 above)
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:890]: This is really fun. The datalayout class asserts if you ask for the stack alignment when the string didn't specify one (which nvptx often doesn't), and has no means of testing directly whether the stack alignment is specified. Hence the TODO below this line. 1024 is a credible guess at probably higher alignment than the stack has naturally. I need a separate patch changing datalayout then rebase this on that one

File: clang/lib/CodeGen/Targets/AMDGPU.cpp (hunk 5 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:121]: Nope. At least, not what I got under testing. That sounds like a good idea though.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 8 above)
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:191]: Oh right. This needs a whole comment block on why this collection of stores is known to create a va_list which is correct on x64 since that's not very obvious. I'll think about how to express that and probably write down the aarch64 one as well, it's the same sort of construct.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 7 above)
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:82]: Written to be consistent with Module::getFunction which uses nullopt. Need to open an issue about this, can't change getDeclaration to return null on missing without causing a lot of breakage. I think the right thing is to rename getDeclaration to getOrCreateDeclaration, introduce another function called getPreexistingDeclaration, wait for at least some of the downstream projects to do the rename, then rename getPreexistingDeclaration into getDeclaration and thus be consistent with Module...

**Comment by JonChesterfield** [comment]
Drive by fixes to some of Matt's comments. Caught a missing line in a .def file for NVPTX through luck due to the enum rename which means the Other/new-pm-thinlto-postlink-samplepgo-defaults.ll style tests need to be patched again - leaving that for now as I want to check whether the passes run in the order of the PassRegistry.def (in which case expand-variadics should probably happen before always-inline) and the failing test is a reminder to myself of that

File: clang/lib/CodeGen/ABIInfoImpl.cpp (hunk 4 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/ABIInfoImpl.cpp]: Yeah, swapping it out for ptr->gettype is a much better choice. That works exactly as one would hope

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 17 above)
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:889]: I think adding the following to DataLayout is reasonable, reflects the existing call to exceeds
```
/// Returns true if the given alignment is below the natural stack alignment.
bool belowNaturalStackAlignment(Align Alignment) const {
  return StackNaturalAlign && (Alignment < *StackNaturalAlign);
}
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 19]
```diff
@@ -0,0 +1,1026 @@
...
+
+// At present Intrinsic:: has no interface to test if a declaration is in the
+// module without creating one. Inserting a declaration and then testing if it
+// has any uses and then deleting it seems a bad way to do the query.
+// Module implements getFunction() which returns nullptr on missing declaration
+// and getOrInsertFunction which creates one when absent. Intrinsics.h
+// implements getDeclaration which creates one when missing. This should be
+// changed to be consistent with Module()'s naming. Implementing as a local
+// function here in the meantime to decouple from that process.
+Function *getPreexistingDeclaration(Module *M, Intrinsic::ID id,
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:82]:
```suggestion
Function *getPreexistingDeclaration(Module *M, Intrinsic::ID Id,
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 20]
```diff
@@ -0,0 +1,1026 @@
...
+
+  // Whether a valist instance is passed by value or by address
+  // I.e. does it need to be alloca'ed and stored into, or can
+  // it be passed directly in a SSA register
+  virtual bool passedInSSARegister() = 0;
+
+  // The type of a va_list iterator object
+  virtual Type *vaListType(LLVMContext &Ctx) = 0;
+
+  // The type of a va_list as a function argument as lowered by C
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:105]: C?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 21]
```diff
@@ -0,0 +1,1026 @@
...
+// Some use more complicated iterator constructs. Type erase that
+// so the rest of the pass can operation on either.
+// Virtual functions where different targets want different behaviour,
+// normal where all implemented targets presently have the same.
+struct VAListInterface {
+  virtual ~VAListInterface() {}
+
+  // Whether a valist instance is passed by value or by address
+  // I.e. does it need to be alloca'ed and stored into, or can
+  // it be passed directly in a SSA register
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:99]: Comments should be sentences (with full stop), here and elsewhere.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 22]
```diff
@@ -0,0 +1,1026 @@
...
+  virtual Type *vaListType(LLVMContext &Ctx) = 0;
+
+  // The type of a va_list as a function argument as lowered by C
+  virtual Type *vaListParameterType(Module &M) = 0;
+
+  // Initialise an allocated va_list object to point to an already
+  // initialised contiguous memory region.
+  // Return the value to pass as the va_list argument
+  virtual Value *initializeVAList(LLVMContext &Ctx, IRBuilder<> &Builder,
+                                  AllocaInst *, Value * /*buffer*/) = 0;
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:112]: No need for the comment. Name all arguments.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 23]
```diff
@@ -0,0 +1,1026 @@
...
+    return PointerType::getUnqual(Ctx);
+  }
+
+  Type *vaListParameterType(Module &M) override {
+    const DataLayout &DL = M.getDataLayout();
+    return DL.getAllocaPtrType(M.getContext());
+  }
+
+  Value *initializeVAList(LLVMContext &Ctx, IRBuilder<> &Builder,
+                          AllocaInst * /*va_list*/, Value *buffer) override {
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:133]: Again style. Names, and capitalization.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 24]
```diff
@@ -0,0 +1,1026 @@
...
+  if (rewriteABI()) {
+    // No direct calls remain to F, remaining uses are things like address
+    // escaping, modulo errors in this implementation.
+    for (User *U : llvm::make_early_inc_range(F->users()))
+      if (CallBase *CB = dyn_cast<CallBase>(U)) {
+        Value *calledOperand = CB->getCalledOperand();
+        if (F == calledOperand) {
+          report_fatal_error(
+              "ExpandVA abi requires eliminating call uses first");
+        }
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:658]: Is this a glorified assert? 

Style: check variable names here and elsewhere, lots of lower case first letters.

**Review by jdoerfert — COMMENTED** [review]
I just read over this and left some comments.

File: clang/lib/CodeGen/ABIInfoImpl.cpp [hunk 25]
```diff
@@ -157,7 +157,7 @@ llvm::Value *CodeGen::emitRoundPointerUpToAlignment(CodeGenFunction &CGF,
   llvm::Value *RoundUp = CGF.Builder.CreateConstInBoundsGEP1_32(
       CGF.Builder.getInt8Ty(), Ptr, Align.getQuantity() - 1);
   return CGF.Builder.CreateIntrinsic(
-      llvm::Intrinsic::ptrmask, {Ptr->getType(), CGF.IntPtrTy},
+        llvm::Intrinsic::ptrmask, {Ptr->getType(), CGF.IntPtrTy},
```
Comment (arsenm) [inline clang/lib/CodeGen/ABIInfoImpl.cpp:160]: Spurious whitespace change?

File: clang/lib/CodeGen/ABIInfoImpl.cpp [hunk 26]
```diff
@@ -247,7 +247,7 @@ Address CodeGen::emitMergePHI(CodeGenFunction &CGF, Address Addr1,
 
 bool CodeGen::isEmptyField(ASTContext &Context, const FieldDecl *FD,
                            bool AllowArrays, bool AsIfNoUniqueAddr) {
-  if (FD->isUnnamedBitField())
+  if (FD->isUnnamedBitfield())
```
Comment (arsenm) [inline clang/lib/CodeGen/ABIInfoImpl.cpp:250]: Unrelated change pulled in?

File: llvm/lib/Target/AMDGPU/AMDGPUPassRegistry.def [hunk 27]
```diff
@@ -24,6 +24,7 @@ MODULE_PASS("amdgpu-lower-ctor-dtor", AMDGPUCtorDtorLoweringPass())
 MODULE_PASS("amdgpu-lower-module-lds", AMDGPULowerModuleLDSPass(*this))
 MODULE_PASS("amdgpu-printf-runtime-binding", AMDGPUPrintfRuntimeBindingPass())
 MODULE_PASS("amdgpu-unify-metadata", AMDGPUUnifyMetadataPass())
+MODULE_PASS("expand-variadics", ExpandVariadicsPass(ExpandVariadicsMode::Lowering))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUPassRegistry.def:27]: Shouldn't need to list this in every target's PassRegistry, the generic one should be fine

File: clang/lib/CodeGen/Targets/AMDGPU.cpp (hunk 5 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp]: ^structs < 8 bytes are packed into integers, but structs larger than that are spread across multiple parameters in a discarding-padding sense. Better testing revealed that one byte aligning things does not work out robustly, have had to take a different strategy to make all the cases work. However that does mean everything is 4 byte aligned now.

**Comment by JonChesterfield** [comment]
I think the comments here are fed into https://github.com/llvm/llvm-project/pull/93362 successfully, will go through the list again to check.

**Comment by arsenm** [comment]
> I think the comments here are fed into #93362 successfully, will go through the list again to check.

So #93362 is the replacement, and not the sequential next piece? Can we close this one then?
